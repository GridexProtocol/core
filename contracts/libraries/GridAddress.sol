// SPDX-License-Identifier: GPL-2.0-or-later
pragma solidity ^0.8.0;

import "@openzeppelin/contracts/utils/Create2.sol";

library GridAddress {
    bytes32 internal constant GRID_BYTES_CODE_HASH = 0x884a6891a166f885bf6f0a3b330a25e41d1761a5aa091110a229d9a0e34b2c36;

    struct GridKey {
        address token0;
        address token1;
        int24 resolution;
    }

    /// @notice Constructs the grid key for the given parameters
    /// @dev tokenA and tokenB may be passed in, in the order of either token0/token1 or token1/token0
    /// @param tokenA The contract address of either token0 or token1
    /// @param tokenB The contract address of the other token
    /// @param resolution The step size in initialized boundaries for a grid created with a given fee
    /// @return key The grid key to compute the canonical address for the grid
    function gridKey(address tokenA, address tokenB, int24 resolution) internal pure returns (GridKey memory key) {
        if (tokenA > tokenB) (tokenA, tokenB) = (tokenB, tokenA);

        return GridKey(tokenA, tokenB, resolution);
    }

    /// @dev Computes the CREATE2 address for a grid with the given parameters
    /// @param gridFactory The address of the grid factory
    /// @param key The grid key to compute the canonical address for the grid
    /// @return grid The computed address
    function computeAddress(address gridFactory, GridKey memory key) internal pure returns (address grid) {
        require(key.token0 < key.token1);
        return
            Create2.computeAddress(
                keccak256(abi.encode(key.token0, key.token1, key.resolution)),
                GRID_BYTES_CODE_HASH,
                gridFactory
            );
    }
}
