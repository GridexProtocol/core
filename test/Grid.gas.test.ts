import {deployERC20, deployGridFactory, deployGridTestHelper, deployWETH} from "./shared/deployer";
import {Resolution, RESOLUTION_X96} from "./shared/util";
import {ethers} from "hardhat";
import {computeAddress, sortedToken} from "./shared/GridAddress";
import {loadFixture} from "@nomicfoundation/hardhat-network-helpers";
import {expect} from "./shared/expect";

describe("Grid", () => {
    const deployFixture = async () => {
        const weth = await deployWETH();
        const {gridFactory} = await deployGridFactory(weth.address);

        const usdc = await deployERC20("USDC", "USDC", 6, 10n ** 18n * 10000n);
        const usdt = await deployERC20("USDT", "USDT", 6, 10n ** 18n * 10000n);

        const gridTestHelper = await deployGridTestHelper(gridFactory.address, weth.address);

        await Promise.all([
            weth.approve(gridTestHelper.address, 10n ** 18n * 10000n),
            usdc.approve(gridTestHelper.address, 10n ** 18n * 10000n),
            usdt.approve(gridTestHelper.address, 10n ** 18n * 10000n),
        ]);

        await gridFactory.createGrid(weth.address, usdc.address, Resolution.MEDIUM);

        await gridFactory.createGrid(usdc.address, usdt.address, Resolution.MEDIUM);

        const ETHToERC20Grid = await ethers.getContractAt(
            "Grid",
            await computeAddress(gridFactory.address, weth.address, usdc.address, Resolution.MEDIUM)
        );

        const [signer] = await ethers.getSigners();
        const parameters = {
            tokenA: weth.address,
            tokenB: usdc.address,
            resolution: Resolution.MEDIUM,
            recipient: signer.address,
            priceX96: RESOLUTION_X96,
            orders0: [
                {
                    boundaryLower: 0,
                    amount: 1n,
                },
            ],
            orders1: [
                {
                    boundaryLower: 0n,
                    amount: 1n,
                },
            ],
        };
        await gridTestHelper.initialize(parameters, {value: 1n});
        await ETHToERC20Grid.settleMakerOrderAndCollectInBatch(signer.address, [1, 2], true);

        const ERC20ToERC20Grid = await ethers.getContractAt(
            "Grid",
            await computeAddress(gridFactory.address, usdc.address, usdt.address, Resolution.MEDIUM)
        );

        parameters.tokenA = usdc.address;
        parameters.tokenB = usdt.address;
        await gridTestHelper.initialize(parameters);
        await ERC20ToERC20Grid.settleMakerOrderAndCollectInBatch(signer.address, [1, 2], false);

        return {weth, usdc, usdt, gridFactory, gridTestHelper, ETHToERC20Grid, ERC20ToERC20Grid};
    };

    it("#createGrid", async () => {
        const {usdc, usdt, gridFactory} = await loadFixture(deployFixture);

        const tx = await gridFactory.createGrid(usdc.address, usdt.address, Resolution.HIGH);
        const receipt = await tx.wait();
        expect(receipt.gasUsed.toNumber()).toMatchSnapshot();
    });

    it("#initialize", async () => {
        const {usdc, usdt, gridFactory, gridTestHelper} = await loadFixture(deployFixture);

        await gridFactory.createGrid(usdc.address, usdt.address, Resolution.HIGH);

        const grid = await ethers.getContractAt(
            "Grid",
            await computeAddress(gridFactory.address, usdc.address, usdt.address, Resolution.HIGH)
        );

        await gridTestHelper.initialize({
            tokenA: usdc.address,
            tokenB: usdt.address,
            resolution: Resolution.HIGH,
            recipient: ethers.constants.AddressZero,
            priceX96: RESOLUTION_X96,
            orders0: [
                {
                    boundaryLower: 0n,
                    amount: 1n,
                },
            ],
            orders1: [
                {
                    boundaryLower: 0n,
                    amount: 1n,
                },
            ],
        });
        expect((await gridTestHelper.gasUsed()).toNumber()).toMatchSnapshot();
    });

    describe("#placeMakerOrder", () => {
        it("ETH", async () => {
            const {weth, usdc, gridTestHelper} = await loadFixture(deployFixture);

            const {token0} = await sortedToken(weth.address, usdc.address);
            for (let i = 0; i < 5; i++) {
                await gridTestHelper.placeMakerOrder(
                    {
                        tokenA: weth.address,
                        tokenB: usdc.address,
                        resolution: Resolution.MEDIUM,
                        recipient: ethers.constants.AddressZero,
                        zero: weth.address.toLowerCase() == token0.toLowerCase(),
                        boundaryLower: 0,
                        amount: 10n ** 18n,
                    },
                    {value: 10n ** 18n}
                );

                expect((await gridTestHelper.gasUsed()).toNumber()).toMatchSnapshot();
            }
        });

        it("ERC20", async () => {
            const {weth, usdc, gridTestHelper} = await loadFixture(deployFixture);

            const {token0} = await sortedToken(weth.address, usdc.address);
            for (let i = 0; i < 5; i++) {
                await gridTestHelper.placeMakerOrder({
                    tokenA: weth.address,
                    tokenB: usdc.address,
                    resolution: Resolution.MEDIUM,
                    recipient: ethers.constants.AddressZero,
                    zero: weth.address.toLowerCase() != token0.toLowerCase(),
                    boundaryLower: 0,
                    amount: 10n ** 18n,
                });
                expect((await gridTestHelper.gasUsed()).toNumber()).toMatchSnapshot();
            }
        });
    });

    describe("#swap", () => {
        describe("exactOutput", () => {
            it("ETH", async () => {
                const {weth, usdc, gridTestHelper} = await loadFixture(deployFixture);

                const {token0} = await sortedToken(weth.address, usdc.address);

                await gridTestHelper.placeMakerOrderInBatch({
                    tokenA: weth.address,
                    tokenB: usdc.address,
                    resolution: Resolution.MEDIUM,
                    recipient: ethers.constants.AddressZero,
                    zero: weth.address.toLowerCase() != token0.toLowerCase(),
                    orders: [
                        {
                            boundaryLower: -Resolution.MEDIUM,
                            amount: 10n ** 18n * 10n,
                        },
                        {
                            boundaryLower: 0,
                            amount: 10n ** 18n * 10n,
                        },
                        {
                            boundaryLower: Resolution.MEDIUM,
                            amount: 10n ** 18n * 10n,
                        },
                    ],
                });

                const signers = await ethers.getSigners();
                for (let i = 0; i < 6; i++) {
                    await gridTestHelper.exactOutput(
                        {
                            tokenIn: weth.address,
                            tokenOut: usdc.address,
                            resolution: Resolution.MEDIUM,
                            recipient: signers[0].address,
                            amountOut: 10n ** 18n,
                            amountInMaximum: 10n ** 18n * 2n,
                            priceLimitX96: 0n,
                        },
                        {value: 10n ** 18n * 2n}
                    );
                    expect((await gridTestHelper.gasUsed()).toNumber()).toMatchSnapshot();
                }
            });

            it("ERC20", async () => {
                const {weth, usdc, gridTestHelper} = await loadFixture(deployFixture);

                const {token0} = await sortedToken(weth.address, usdc.address);

                await gridTestHelper.placeMakerOrderInBatch(
                    {
                        tokenA: weth.address,
                        tokenB: usdc.address,
                        resolution: Resolution.MEDIUM,
                        recipient: ethers.constants.AddressZero,
                        zero: weth.address.toLowerCase() == token0.toLowerCase(),
                        orders: [
                            {
                                boundaryLower: -Resolution.MEDIUM,
                                amount: 10n ** 18n * 10n,
                            },
                            {
                                boundaryLower: 0,
                                amount: 10n ** 18n * 10n,
                            },
                            {
                                boundaryLower: Resolution.MEDIUM,
                                amount: 10n ** 18n * 10n,
                            },
                        ],
                    },
                    {value: 10n ** 18n * 10n * 3n}
                );

                const signers = await ethers.getSigners();
                for (let i = 0; i < 6; i++) {
                    await gridTestHelper.exactOutput({
                        tokenIn: usdc.address,
                        tokenOut: weth.address,
                        resolution: Resolution.MEDIUM,
                        recipient: signers[0].address,
                        amountOut: 10n ** 18n,
                        amountInMaximum: 10n ** 18n * 2n,
                        priceLimitX96: 0n,
                    });
                    expect((await gridTestHelper.gasUsed()).toNumber()).toMatchSnapshot();
                }
            });
        });

        describe("exactInput", () => {
            it("ETH", async () => {
                const {weth, usdc, gridTestHelper} = await loadFixture(deployFixture);

                const {token0} = await sortedToken(weth.address, usdc.address);

                await gridTestHelper.placeMakerOrderInBatch({
                    tokenA: weth.address,
                    tokenB: usdc.address,
                    resolution: Resolution.MEDIUM,
                    recipient: ethers.constants.AddressZero,
                    zero: weth.address.toLowerCase() != token0.toLowerCase(),
                    orders: [
                        {
                            boundaryLower: -Resolution.MEDIUM,
                            amount: 10n ** 18n * 10n,
                        },
                        {
                            boundaryLower: 0,
                            amount: 10n ** 18n * 10n,
                        },
                        {
                            boundaryLower: Resolution.MEDIUM,
                            amount: 10n ** 18n * 10n,
                        },
                    ],
                });

                const signers = await ethers.getSigners();
                for (let i = 0; i < 6; i++) {
                    await gridTestHelper.exactInput(
                        {
                            tokenIn: weth.address,
                            tokenOut: usdc.address,
                            resolution: Resolution.MEDIUM,
                            recipient: signers[0].address,
                            amountIn: 10n ** 18n / 2n,
                            amountOutMinimum: 0n,
                            priceLimitX96: 0n,
                        },
                        {value: 10n ** 18n / 2n}
                    );
                    expect((await gridTestHelper.gasUsed()).toNumber()).toMatchSnapshot();
                }
            });

            it("ERC20", async () => {
                const {weth, usdc, gridTestHelper} = await loadFixture(deployFixture);

                const {token0} = await sortedToken(weth.address, usdc.address);

                await gridTestHelper.placeMakerOrderInBatch(
                    {
                        tokenA: weth.address,
                        tokenB: usdc.address,
                        resolution: Resolution.MEDIUM,
                        recipient: ethers.constants.AddressZero,
                        zero: weth.address.toLowerCase() == token0.toLowerCase(),
                        orders: [
                            {
                                boundaryLower: -Resolution.MEDIUM,
                                amount: 10n ** 18n * 10n,
                            },
                            {
                                boundaryLower: 0,
                                amount: 10n ** 18n * 10n,
                            },
                            {
                                boundaryLower: Resolution.MEDIUM,
                                amount: 10n ** 18n * 10n,
                            },
                        ],
                    },
                    {value: 10n ** 18n * 10n * 3n}
                );

                const signers = await ethers.getSigners();
                for (let i = 0; i < 6; i++) {
                    await gridTestHelper.exactInput({
                        tokenIn: usdc.address,
                        tokenOut: weth.address,
                        resolution: Resolution.MEDIUM,
                        recipient: signers[0].address,
                        amountIn: 10n ** 18n / 2n,
                        amountOutMinimum: 0n,
                        priceLimitX96: 0n,
                    });
                    expect((await gridTestHelper.gasUsed()).toNumber()).toMatchSnapshot();
                }
            });
        });
    });
});
