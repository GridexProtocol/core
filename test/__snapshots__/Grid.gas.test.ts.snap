// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid #createGrid 1`] = `6698932`;

exports[`Grid #initialize 1`] = `501599`;

exports[`Grid #placeMakerOrder ERC20 1`] = `168571`;

exports[`Grid #placeMakerOrder ERC20 2`] = `111701`;

exports[`Grid #placeMakerOrder ERC20 3`] = `111701`;

exports[`Grid #placeMakerOrder ERC20 4`] = `111701`;

exports[`Grid #placeMakerOrder ERC20 5`] = `111701`;

exports[`Grid #placeMakerOrder ETH 1`] = `203225`;

exports[`Grid #placeMakerOrder ETH 2`] = `129244`;

exports[`Grid #placeMakerOrder ETH 3`] = `129244`;

exports[`Grid #placeMakerOrder ETH 4`] = `129244`;

exports[`Grid #placeMakerOrder ETH 5`] = `129244`;

exports[`Grid #swap exactInput ERC20 1`] = `162455`;

exports[`Grid #swap exactInput ERC20 2`] = `91970`;

exports[`Grid #swap exactInput ERC20 3`] = `91970`;

exports[`Grid #swap exactInput ERC20 4`] = `91953`;

exports[`Grid #swap exactInput ERC20 5`] = `104232`;

exports[`Grid #swap exactInput ERC20 6`] = `91952`;

exports[`Grid #swap exactInput ETH 1`] = `144205`;

exports[`Grid #swap exactInput ETH 2`] = `110024`;

exports[`Grid #swap exactInput ETH 3`] = `110024`;

exports[`Grid #swap exactInput ETH 4`] = `110024`;

exports[`Grid #swap exactInput ETH 5`] = `122305`;

exports[`Grid #swap exactInput ETH 6`] = `110054`;

exports[`Grid #swap exactOutput ERC20 1`] = `162300`;

exports[`Grid #swap exactOutput ERC20 2`] = `91798`;

exports[`Grid #swap exactOutput ERC20 3`] = `104077`;

exports[`Grid #swap exactOutput ERC20 4`] = `91828`;

exports[`Grid #swap exactOutput ERC20 5`] = `104107`;

exports[`Grid #swap exactOutput ERC20 6`] = `91812`;

exports[`Grid #swap exactOutput ETH 1`] = `143598`;

exports[`Grid #swap exactOutput ETH 2`] = `121679`;

exports[`Grid #swap exactOutput ETH 3`] = `109428`;

exports[`Grid #swap exactOutput ETH 4`] = `121634`;

exports[`Grid #swap exactOutput ETH 5`] = `108537`;

exports[`Grid #swap exactOutput ETH 6`] = `121618`;
